# Keras Exercise

## Predict political party based on votes

As a fun little example, we'll use a public data set of how US congressmen voted on 17 different issues in the year 1984. Let's see if we can figure out their political party based on their votes alone, using a deep neural network!

For those outside the United States, our two main political parties are "Democrat" and "Republican." In modern times they represent progressive and conservative ideologies, respectively.

Politics in 1984 weren't quite as polarized as they are today, but you should still be able to get over 90% accuracy without much trouble.

Since the point of this exercise is implementing neural networks in Keras, I'll help you to load and prepare the data.

Let's start by importing the raw CSV file using Pandas, and make a DataFrame out of it with nice column labels:

In [12]:
import pandas as pd

feature_names =  ['party','handicapped-infants', 'water-project-cost-sharing', 
                    'adoption-of-the-budget-resolution', 'physician-fee-freeze',
                    'el-salvador-aid', 'religious-groups-in-schools',
                    'anti-satellite-test-ban', 'aid-to-nicaraguan-contras',
                    'mx-missle', 'immigration', 'synfuels-corporation-cutback',
                    'education-spending', 'superfund-right-to-sue', 'crime',
                    'duty-free-exports', 'export-administration-act-south-africa']

voting_data = pd.read_csv('house-votes-84.data.txt', na_values=['?'], 
                          names = feature_names)
voting_data.head()

party handicapped-infants water-project-cost-sharing  \
0  republican                   n                          y   
1  republican                   n                          y   
2    democrat                 NaN                          y   
3    democrat                   n                          y   
4    democrat                   y                          y   

  adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
0                                 n                    y               y   
1                                 n                    y               y   
2                                 y                  NaN               y   
3                                 y                    n             NaN   
4                                 y                    n               y   

  religious-groups-in-schools anti-satellite-test-ban  \
0                           y                       n   
1                           y                       n   
2                           y                       n   
3                           y                       n   
4                           y                       n   

  aid-to-nicaraguan-contras mx-missle immigration  \
0                         n         n           y   
1                         n         n           n   
2                         n         n           n   
3                         n         n           n   
4                         n         n           n   

  synfuels-corporation-cutback education-spending superfund-right-to-sue  \
0                          NaN                  y                      y   
1                            n                  y                      y   
2                            y                  n                      y   
3                            y                  n                      y   
4                            y                NaN                      y   

  crime duty-free-exports export-administration-act-south-africa  
0     y                 n                                      y  
1     y                 n                                    NaN  
2     y                 n                                      n  
3     n                 n                                      y  
4     y                 y                                      y

We can use describe() to get a feel of how the data looks in aggregate:

In [14]:
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        435                 423                        387   
unique         2                   2                          2   
top     democrat                   n                          y   
freq         267                 236                        195   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                424                  424             420   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 253                  247             212   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          424                     421   
unique                           2                       2   
top                              y                       y   
freq                           272                     239   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        420       413         428   
unique                         2         2           2   
top                            y         y           y   
freq                         242       207         216   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           414                404                    410   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            264                233                    209   

       crime duty-free-exports export-administration-act-south-africa  
count    418               407                                    331  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     248               233                                    269

We can see there's some missing data to deal with here; some politicians abstained on some votes, or just weren't present when the vote was taken. We will just drop the rows with missing data to keep it simple, but in practice you'd want to first make sure that doing so didn't introduce any sort of bias into your analysis (if one party abstains more than another, that could be problematic for example.)

In [15]:
voting_data.dropna(inplace=True)
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        232                 232                        232   
unique         2                   2                          2   
top     democrat                   n                          n   
freq         124                 136                        125   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                232                  232             232   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 123                  119             128   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          232                     232   
unique                           2                       2   
top                              y                       y   
freq                           149                     124   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        232       232         232   
unique                         2         2           2   
top                            y         n           y   
freq                         119       119         128   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           232                232                    232   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            152                124                    127   

       crime duty-free-exports export-administration-act-south-africa  
count    232               232                                    232  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     149               146                                    189

Our neural network needs normalized numbers, not strings, to work. So let's replace all the y's and n's with 1's and 0's, and represent the parties as 1's and 0's as well.

In [20]:
voting_data.replace(('y', 'n'), (1, 0), inplace=True)
voting_data.replace(('democrat', 'republican'), (1, 0), inplace=True)

In [21]:
voting_data.head()

party  handicapped-infants  water-project-cost-sharing  \
5       1                    0                           1   
8       0                    0                           1   
19      1                    1                           1   
23      1                    1                           1   
25      1                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

Finally let's extract the features and labels in the form that Keras will expect:

In [22]:
all_features = voting_data[feature_names].drop('party', axis=1).values
all_classes = voting_data['party'].values

OK, so have a go at it! You'll want to refer back to the slide on using Keras with binary classification - there are only two parties, so this is a binary problem. This also saves us the hassle of representing classes with "one-hot" format like we had to do with MNIST; our output is just a single 0 or 1 value.

Also refer to the scikit_learn integration slide, and use cross_val_score to evaluate your resulting model with 10-fold cross-validation.

**If you're using tensorflow-gpu on a Windows machine** by the way, you probably *do* want to peek a little bit at my solution - if you run into memory allocation errors, there's a workaround there you can use.

Try out your code here:

In [33]:
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score


def ai_model():
    # Binary Classification
    model = Sequential()

    #16 feature inputs (votes) going into an 32-unit layer 
    model.add(Dense(32, input_dim=16, kernel_initializer='normal', activation='relu'))
    # Dropout to prevent overfiting
    model.add(Dropout(0.5))

    # Hidden Layer 1
    model.add(Dense(32, kernel_initializer='normal', activation='relu'))
    # Dropout to prevent overfiting
    model.add(Dropout(0.5))
    
    # Hidden Layer 2
    model.add(Dense(32, kernel_initializer='normal', activation='relu'))
    # Dropout to prevent overfiting
    model.add(Dropout(0.5))
    
    # Hidden Layer 3
    model.add(Dense(32, kernel_initializer='normal', activation='relu'))
    # Dropout to prevent overfiting
    model.add(Dropout(0.5))
    
    # Hidden Layer 4
    model.add(Dense(16, kernel_initializer='normal', activation='relu'))
    # Dropout to prevent overfiting
    model.add(Dropout(0.5))

    # Output Layer with binary classification: "Democrat" or "Republican"
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# Wrap our Keras model in an estimator compatible with scikit_learn
estimator = KerasClassifier(build_fn=ai_model, epochs=100, verbose=0)

# Number of steps
n = 10

# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features, all_classes, cv=n)
cv_scores.mean()

/tmp/ipykernel_124077/2779962516.py:42: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  estimator = KerasClassifier(build_fn=ai_model, epochs=100, verbose=0)


0.9525362312793731

## My implementation is below

# No peeking!

![title](peek.jpg)

In [7]:
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score

def create_model():
    model = Sequential()
    #16 feature inputs (votes) going into an 32-unit layer 
    model.add(Dense(32, input_dim=16, kernel_initializer='normal', activation='relu'))
    # Another hidden layer of 16 units
    model.add(Dense(16, kernel_initializer='normal', activation='relu'))
    # Output layer with a binary classification (Democrat or Republican political party)
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

# Wrap our Keras model in an estimator compatible with scikit_learn
estimator = KerasClassifier(build_fn=create_model, epochs=100, verbose=0)
# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features, all_classes, cv=10)
cv_scores.mean()

2023-02-02 17:37:38.856550: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-02 17:37:38.959831: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-02-02 17:37:38.959850: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-02-02 17:37:39.545327: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2023-

0.9438405811786652

94% without even trying too hard! Did you do better? Maybe more neurons, more layers, or Dropout layers would help even more.